tracer/src/Datadog.Trace/Activity/ActivityListener.cs

```diff
@@ -0,0 +1,511 @@
+
+            DuckType.CreateTypeResult onNextActivityProxyResult;
+            if (_activityType.GetField("_traceId", DuckAttribute.DefaultFlags) is not null)
+            {
+                onNextActivityProxyResult = DuckType.GetOrCreateProxyType(typeof(IW3CActivity), _activityType);
+            }
+            else
+            {
+                onNextActivityProxyResult = DuckType.GetOrCreateProxyType(typeof(IActivity), _activityType);
+            }
+
+            var onNextActivityMethod = _onNextActivityMethodInfo.MakeGenericMethod(onNextActivityProxyResult.ProxyType);
+            var onNextActivityProxyTypeCtor = onNextActivityProxyResult.ProxyType.GetConstructors()[0];
+            var onNextActivityDynMethodIl = onNextActivityDynMethod.GetILGenerator();
+            onNextActivityDynMethodIl.Emit(OpCodes.Ldarg_0);
+            onNextActivityDynMethodIl.Emit(OpCodes.Ldarg_1);
+            onNextActivityDynMethodIl.Emit(OpCodes.Ldarg_2);
+            onNextActivityDynMethodIl.Emit(OpCodes.Newobj, onNextActivityProxyTypeCtor);
+            onNextActivityDynMethodIl.EmitCall(OpCodes.Call, onNextActivityMethod, null);
+            onNextActivityDynMethodIl.Emit(OpCodes.Ret);
+            _onNextActivityDelegate = (Action<string, KeyValuePair<string, object>, object>)onNextActivityDynMethod.CreateDelegate(typeof(Action<string, KeyValuePair<string, object>, object>));
+
+            // Create delegate for OnSetListener + Source
+            var onSetListenerDynMethod = new DynamicMethod("OnShouldListenToDyn", typeof(bool), new[] { typeof(object) }, typeof(ActivityListener).Module, true);
+            var onSetListenerProxyResult = DuckType.GetOrCreateProxyType(typeof(ISource), _diagnosticListenerType);
+            var onSetListenerMethod = _onShouldListenToMethodInfo.MakeGenericMethod(onSetListenerProxyResult.ProxyType);
+            var onSetListenerProxyTypeCtor = onSetListenerProxyResult.ProxyType.GetConstructors()[0];
+            var onSetListenerDynMethodIl = onSetListenerDynMethod.GetILGenerator();
+            onSetListenerDynMethodIl.Emit(OpCodes.Ldarg_0);
+            onSetListenerDynMethodIl.Emit(OpCodes.Newobj, onSetListenerProxyTypeCtor);
+            onSetListenerDynMethodIl.EmitCall(OpCodes.Call, onSetListenerMethod, null);
+            onSetListenerDynMethodIl.Emit(OpCodes.Ret);
+            _onSetListenerDelegate = (Func<object, bool>)onSetListenerDynMethod.CreateDelegate(typeof(Func<object, bool>));
+
+            // Initialize and subscribe to DiagnosticListener.AllListeners.Subscribe
+            var diagObserverType = CreateDiagnosticObserverType();
+            var diagListener = Activator.CreateInstance(diagObserverType);
+            var allListenersPropertyInfo = _diagnosticListenerType.GetProperty("AllListeners", BindingFlags.Public | BindingFlags.Static);
+            var subscribeMethodInfo = allListenersPropertyInfo?.PropertyType.GetMethod("Subscribe");
```

Comment:
To be safe, it might be worth specifying the parameters array in the GetMethod call to make sure we get the overload that only has the one `IObserver<KeyValuePair<string,object>>` parameter

Reply:
done in https://github.com/DataDog/dd-trace-dotnet/pull/2446/commits/2a3e10e1038eb27932fccc754f50b100cee1dd22